### PR TITLE
Improved build logging

### PR DESCRIPTION
**Pull request by niloc132**
This patch makes the BuildLog interface actually useful, shaped mostly like Maven's own log API, and plumbs it not only to send some basic progress details about the build itself, but also collects errors/warnings/notes about each build. Those are printed now into the main build log, and also written to disk so they can be reviewed later, or possibly repeated for cached results.

TaskOutput was renamed to TaskContext, and implements the BuildLog api. In the future this repurposed class will probably also be how a task gets access to incremental build details, if any.

### CODE REVIEW

**Comment by niloc132** [comment]
Pjng @xamde if you want to make sure this still has the spirit you were after, and I didn't make any obvious mistakes along the way.

File: build-caching/src/main/java/com/vertispan/j2cl/build/TaskScheduler.java [hunk 1]
```diff
@@ -294,7 +304,6 @@ private boolean executeFinalTask(CollectedTaskInputs taskDetails, DiskCache.Cach
                             throw new AssertionError("final task marker should have been " + cacheResult.outputDir() + ", instead was " + previous);
                         }
                     }
-                    buildLog.info("--- Ready for browser refresh");
```
Comment (xamde) [inline build-caching/src/main/java/com/vertispan/j2cl/build/TaskScheduler.java]: Having a clear signal on the terminal when the relevant step has been done turned out to be really helpful. Every tasks prints "Finished" something so one needs to read carefully.

**Review by xamde — APPROVED** [review]
Wow, this looks good. You clearly 10x-ed my humble BuildLog beginnings, thanks!

**Comment by niloc132** [comment]
Thank you for taking the initiative and getting it started.

**Pull request merged**